File: scripts/scrape-history.js
'use strict';
// Run once to generate data/history.json
// Usage: node scripts/scrape-history.js
// Takes ~30 minutes (500ms delay × 366 days).

const https = require('https');
const fs    = require('fs');
const path  = require('path');

const MONTHS = [
  'January','February','March','April','May','June',
  'July','August','September','October','November','December'
];
const DAYS_IN_MONTH = [31,29,31,30,31,30,31,31,30,31,30,31];

function get(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, {
      headers: {
        'User-Agent': 'EpochInstituteHistoryScraper/1.0 (build script; research@the-epoch-institute.org)',
        'Accept': 'application/json'
      }
    }, res => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return get(res.headers.location).then(resolve).catch(reject);
      }
      let body = '';
      res.on('data', chunk => body += chunk);
      res.on('end', () => resolve({ status: res.statusCode, body }));
    });
    req.on('error', reject);
    req.setTimeout(15000, () => { req.destroy(); reject(new Error('timeout')); });
  });
}

function delay(ms) {
  return new Promise(r => setTimeout(r, ms));
}

function cleanText(t) {
  return (t || '')
    .replace(/\[[\d\w,\s]+\]/g, '')  // remove [1], [a], [note 1], etc.
    .replace(/\s+/g, ' ')
    .trim();
}

function makeTitle(text) {
  let t = text.replace(/^(The |A |An )/i, '');
  // Cut at first comma, em dash, or en dash
  const cut = t.search(/,|\s[—–]\s/);
  if (cut > 4 && cut < 80) t = t.slice(0, cut).trim();
  // Limit to 9 words
  const words = t.split(/\s+/);
  if (words.length > 9) t = words.slice(0, 9).join(' ');
  return t.trim().replace(/[,;.]$/, '');
}

function makeDescription(text) {
  const clean = cleanText(text);
  // Take first sentence
  const m = clean.match(/^(.+?[.!?])(?:\s|$)/);
  let desc = m ? m[1] : clean;
  // Truncate if too long (> 22 words)
  const words = desc.split(/\s+/);
  if (words.length > 22) desc = words.slice(0, 20).join(' ') + '.';
  return desc.trim();
}

function formatYearDisplay(year) {
  if (typeof year !== 'number') return String(year);
  return year < 0 ? `${Math.abs(year)} BCE` : String(year);
}

async function scrapeDay(month, day) {
  const url = `https://en.wikipedia.org/api/rest_v1/feed/onthisday/events/${month}/${day}`;
  try {
    const { status, body } = await get(url);
    if (status !== 200) {
      console.warn(`  HTTP ${status} for ${MONTHS[month - 1]} ${day}`);
      return [];
    }
    const data = JSON.parse(body);
    const events = Array.isArray(data.events) ? data.events : [];
    return events
      .filter(ev => typeof ev.year === 'number' && ev.text)
      .map(ev => {
        const cleaned = cleanText(ev.text);
        return {
          year:        ev.year,
          yearDisplay: formatYearDisplay(ev.year),
          title:       makeTitle(cleaned),
          description: makeDescription(ev.text)
        };
      })
      .filter(ev => ev.title.length > 3 && ev.description.length > 10)
      .sort((a, b) => a.year - b.year);
  } catch (err) {
    console.warn(`  Error on ${MONTHS[month - 1]} ${day}: ${err.message}`);
    return [];
  }
}

async function main() {
  const output = {};
  let total = 0;
  const start = Date.now();

  for (let m = 1; m <= 12; m++) {
    for (let d = 1; d <= DAYS_IN_MONTH[m - 1]; d++) {
      const key    = String(m).padStart(2, '0') + '-' + String(d).padStart(2, '0');
      const events = await scrapeDay(m, d);
      output[key]  = events;
      total       += events.length;
      console.log(`Scraped ${MONTHS[m - 1]} ${String(d).padStart(2)} — ${events.length} events`);
      await delay(500);
    }
  }

  const outDir  = path.resolve(__dirname, '..', 'data');
  fs.mkdirSync(outDir, { recursive: true });
  const outPath = path.join(outDir, 'history.json');
  fs.writeFileSync(outPath, JSON.stringify(output, null, 2), 'utf8');

  const mins = ((Date.now() - start) / 60000).toFixed(1);
  console.log(`\nDone — ${total} events across 366 days in ${mins} min`);
  console.log(`Saved to ${outPath}`);
  console.log(`\nNext steps:`);
  console.log(`  git add data/history.json`);
  console.log(`  git commit -m "Add scraped history data"`);
  console.log(`  git push && deploy via cPanel`);
}

main().catch(err => { console.error(err); process.exit(1); });

File: scripts/scrape-history.py
"""
One-time build script. Generates data/history.json from Wikipedia's On This Day API.
Run: python scripts/scrape-history.py
Takes ~30 minutes (0.5s delay x 366 days).
"""
import json, re, time, urllib.request, urllib.error, os, sys

MONTHS = [
    'January','February','March','April','May','June',
    'July','August','September','October','November','December'
]
DAYS_IN_MONTH = [31,29,31,30,31,30,31,31,30,31,30,31]

HEADERS = {
    'User-Agent': 'EpochInstituteHistoryScraper/1.0 (build script; research@the-epoch-institute.org)',
    'Accept': 'application/json'
}

def fetch(url, retries=3):
    for attempt in range(retries):
        try:
            req = urllib.request.Request(url, headers=HEADERS)
            with urllib.request.urlopen(req, timeout=20) as r:
                return r.read().decode('utf-8')
        except Exception as e:
            if attempt < retries - 1:
                time.sleep(2)
            else:
                raise
    return None

def clean_text(t):
    t = re.sub(r'\[[\w\d\s,]+\]', '', t or '')
    return re.sub(r'\s+', ' ', t).strip()

def make_title(text):
    t = re.sub(r'^(The |A |An )', '', text, flags=re.IGNORECASE)
    m = re.search(r',|\s[—–]\s', t)
    if m and 4 < m.start() < 80:
        t = t[:m.start()].strip()
    words = t.split()
    if len(words) > 9:
        t = ' '.join(words[:9])
    return t.strip().rstrip('.,;')

def make_description(text):
    cleaned = clean_text(text)
    m = re.match(r'^(.+?[.!?])(?:\s|$)', cleaned)
    desc = m.group(1) if m else cleaned
    words = desc.split()
    if len(words) > 22:
        desc = ' '.join(words[:20]) + '.'
    return desc.strip()

def format_year(year):
    if isinstance(year, int):
        return f'{abs(year)} BCE' if year < 0 else str(year)
    return str(year)

def scrape_day(month, day):
    url = f'https://en.wikipedia.org/api/rest_v1/feed/onthisday/events/{month}/{day}'
    try:
        body = fetch(url)
        data = json.loads(body)
        events = data.get('events', [])
        result = []
        for ev in events:
            if not isinstance(ev.get('year'), int) or not ev.get('text'):
                continue
            cleaned = clean_text(ev['text'])
            title = make_title(cleaned)
            desc  = make_description(ev['text'])
            if len(title) < 4 or len(desc) < 10:
                continue
            result.append({
                'year':        ev['year'],
                'yearDisplay': format_year(ev['year']),
                'title':       title,
                'description': desc
            })
        result.sort(key=lambda e: e['year'])
        return result
    except Exception as e:
        print(f'  Error on {MONTHS[month-1]} {day}: {e}', file=sys.stderr)
        return []

def main():
    output = {}
    total  = 0
    start  = time.time()

    for m in range(1, 13):
        for d in range(1, DAYS_IN_MONTH[m-1] + 1):
            key    = f'{m:02d}-{d:02d}'
            events = scrape_day(m, d)
            output[key] = events
            total += len(events)
            print(f'Scraped {MONTHS[m-1]:>10} {d:2d}  —  {len(events)} events')
            sys.stdout.flush()
            time.sleep(0.5)

    out_dir  = os.path.join(os.path.dirname(__file__), '..', 'data')
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, 'history.json')
    with open(out_path, 'w', encoding='utf-8') as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    elapsed = (time.time() - start) / 60
    print(f'\nDone — {total} events across 366 days in {elapsed:.1f} min')
    print(f'Saved to {os.path.abspath(out_path)}')

if __name__ == '__main__':
    main()
